tcms/management/tests/test_admin.py

```diff
@@ -0,0 +1,37 @@
+from django.urls import reverse
+
+from tcms.tests import LoggedInTestCase
+from tcms.tests.factories import BuildFactory, ComponentFactory
+from tcms.utils.permissions import initiate_user_with_default_setups
+
+
+class TestComponentAdmin(LoggedInTestCase):
+    @classmethod
+    def setUpTestData(cls):
+        super().setUpTestData()
+        initiate_user_with_default_setups(cls.tester)
+        ComponentFactory()
+
+    def test_generated_database_queries(self):
+        self.client.login(  # nosec:B106:hardcoded_password_funcarg
+            username=self.tester.username, password="password"
+        )
+        with self.assertNumQueries(12):
```

Comment:
can you explain how this validates that we're using .select_related() on the query set ?

Reply:
Without select_related(), 13 queries are generated and with it only 12.

Reply:
This is not reliable. It will break the moment some of the Django internals make an additional query. Also it is a black-box. A number doesn't tell you anything about the specifics of the query. 

See the internal implementation of `assertNumQueries()` and also see this example:

```
>>> with CaptureQueriesContext(connection) as context:
...     Component.objects.all()
...     print("1: ", context.captured_queries)
... 
<QuerySet [<Component: terminal>, <Component: website>]>
1:  [{'sql': 'SELECT "management_component"."id", "management_component"."name", "management_component"."product_id", "management_component"."initial_owner_id", "management_component"."initial_qa_contact_id", "management_component"."description" FROM "management_component" ORDER BY "management_component"."name" ASC LIMIT 21', 'time': '0.005'}]
>>> 
>>> 
>>> with CaptureQueriesContext(connection) as context:
...     Component.objects.select_related('product', 'initial_owner').all()
...     print("2: ", context.captured_queries)
... 
<QuerySet [<Component: terminal>, <Component: website>]>
2:  [{'sql': 'SELECT "management_component"."id", "management_component"."name", "management_component"."product_id", "management_component"."initial_owner_id", "management_component"."initial_qa_contact_id", "management_component"."description", "management_product"."id", "management_product"."name", "management_product"."classification_id", "management_product"."description", "auth_user"."id", "auth_user"."password", "auth_user"."last_login", "auth_user"."is_superuser", "auth_user"."username", "auth_user"."first_name", "auth_user"."last_name", "auth_user"."email", "auth_user"."is_staff", "auth_user"."is_active", "auth_user"."date_joined" FROM "management_component" INNER JOIN "management_product" ON ("management_component"."product_id" = "management_product"."id") LEFT OUTER JOIN "auth_user" ON ("management_component"."initial_owner_id" = "auth_user"."id") ORDER BY "management_component"."name" ASC LIMIT 21', 'time': '0.003'}]
```

Reply:
@atodorov I have updated the test. Is it correct now?